ipa-core/src/helpers/transport/handler.rs

```diff
@@ -173,18 +173,13 @@ pub enum Error {
 /// Trait for custom-handling different request types made against MPC helper parties.
 /// There is a limitation for RPITIT that traits can't be made object-safe, hence the use of async_trait
 #[async_trait]
-pub trait RequestHandler: Send + Sync {
-    type Identity: TransportIdentity;
+pub trait RequestHandler<I: TransportIdentity>: Send + Sync {
```

Comment:
It was intentional to have `Identity` as an associated type - that ensures for each peer type we have exactly one handler. Unless we have to change it, I would prefer to have it as AT

Reply:
Can you please clarify "ensures for each peer type we have exactly one handler"? The following code is perfectly fine to add:
```
struct AnotherHandler;

#[async_trait]
impl RequestHandler for AnotherHandler {
    type Identity = HelperIdentity;

    async fn handle(
        &self,
        req: Addr<HelperIdentity>,
        data: BodyStream,
    ) -> Result<HelperResponse, ApiError> {
        Ok(HelperResponse::ok())
    }
}
```
In other words, an associated type CAN have multiple handlers.


An associated type prevents having a single implementor support multiple associated types. I don't think that's a desirable property here. More concretely, It's simpler to have `Inner` implement both Shard and Mpc Handlers. The alternative is to have 2 structs which are going to be interconnected anyway because they want to use the same Query Processor and transports. I would advocate on keeping things simple and avoid proliferation of wrappers. That's what I'm proposing on this PR.


Let me know what you think, but I will need some more clarification on what you're overall trying to do to find the right abstraction.

Reply:
The concrete case where it's needed / desirable to implement both kinds of handlers is the `HelperApp` (it wasn't needed previously, but it is needed now, to handle prepare-query messages from the leader shard).

A lot of the uses of `RequestHandler` are via `make_owned_handler`, where there's an indirect limitation that the handler can only handle one identity type, because the handler wraps `Fn(Addr<I>)`.

I'm a bit surprised, but I can't even find other concrete types implementing `RequestHandler` at all, so I can't assess risk of incorrectly implementing `RequestHandler` for multiple identity types. Are there places I'm missing?

Reply:
@andyleiserson We wanted to keep request handlers in one place, so not having too many implementations of `RequestHandlers` was the goal of #992. 

Given that we need a single type to implement both, generic is the way to go